### PR TITLE
Blocky keyboard for #327

### PR DESCRIPTION
**Pull request by zelmon64**
Added a blocky keyboard where there are much fewer keys and the letters are grouped together. The user must select a group (and subgroup) before the desired key can be pressed. This is mainly for when the user needs to use larger than normal keys. This keyboard can only be accessed by setting it as the startup keyboard.

The setting `EnableQuitKeys` has been added to be able to hide the quit keys so that the only way to quit is by right-clicking on the keyboard and selecting Quit.

### CODE REVIEW

**Comment by JuliusSweetland** [comment]
@zelmon64 I've reviewed the PR and have a few changes I would like to make. If I list out my initial thoughts can you tell me what you think and whether you'll have a chance to make them? I am happy to work with you, but I won't be able to get any work done until later this week, or this weekend. It's an excellent PR by the way - these are mainly consistency points to keep the code base as clean as possible. I'm a big believer in consistency so people can start to learn the code base and then guess how other parts work and be correct more often than they are wrong...

1.Rename Blocky to something else, e.g. SimplifiedAlpha. The keyboard could be SimplifiedAlpha and should be located in the English folder as it will need to be localised into other languages if people want to use it. Function keys could be SimplifiedAlphaABC etc. Resources such as BLOCKY_ALPHA_LAYOUT should be renamed also.

2.What do you think about having 2 version of the Simplified keyboard; SimplifiedAlpha and SimplifiedConversationAlpha (in the same way as AlphabeticalAlpha and AlphabeticalConversationAlpha). They could be identical for now with SimplifiedAlpha being used to type into other applications. We could later add a few more features to the SimplifiedAlpha, such as a shift key, but we can discuss this later.

3.I think we need a new setting like the "Use alphabetical keyboard layout" (UseAlphabeticalKeyboardLayout) setting (called UseSimplifiedKeyboardLayout) as the Simplified alpha keyboard is really just an alternative alpha keyboard. This, like the UseAlphabeticalKeyboardLayout setting, should only be available if the keyboard language is English, and impacts the KeyboardHost to return the SimplifiedAlpha keyboard when Alpha is selected, exactly like the UseAlphabeticalKeyboardLayout setting. The Simplified keyboards would therefore not be options in the startup keyboard list, instead you would turn on UseSimplifiedKeyboardLayout ("Use simplify keyboard layout" setting text) and set the startup keyboard to Alpha. If UseSimplifiedKeyboardLayout and UseAlphabeticalKeyboardLayout are both on then UseSimplifiedKeyboardLayout takes precedence.

4.Instead of the very lowest levels of simplified keys being function keys where the handlers directly edit the Text property, they can be normal keys with text, i.e remove BlockyA, BlockyAp, BlockyEx, BlockyPe, etc and just place with keys with Text="A" etc. This impacts a couple of things;

a) Casing. You currently handle this by always outputting uppercase letters, but if the KeyboardOutputService is going to handle the output then we need to force caps locks. This opens the possibility of adding a shift key to the Simplified keyboards to let people control casing, or to use the auto-capitalisation logic that is build in. For now I think if the user turns on the UseSimplifiedKeyboardLayout setting then we should prompt the user to also turn on Settings.Default.ForceCapsLock, and vice versa (prompt user to turn off ForceCapsLock when they turn of UseSimplifiedKeyboardLayout). When UseSimplifiedKeyboardLayout setting is turned on then also prompt the user to change the startup keyboard to ConversationAlpha as that is probably what is desired.

image

b) Need new logic to determine which keyboard to return to after outputting the character. This can be moved into the MainViewModel -  you would need to change the IKeyboardOutputService to implement INotifyPropertyChanged (which it does already) then monitor changes to the Text property - if the .LastOrDefault() character is not null and is a letter (char.IsLetter) then return to SimplifiedAlpha, else if it is char.IsNumber then return to the main SimplifiedNumeric keyboard. Symbols might have to be case by case as they are split over those two keyboards.

4.Change casing on Enable_Quit_Keys_LABEL resource to all uppercase for consistency.

5.Rename BlockyLayer setting to "SimplifiedKeyboardCurrentContext"

**Comment by zelmon64** [comment]
@JuliusSweetland I think these are all good points and I should be able to start on them tomorrow. I especially like the part about making it similar to AlphabeticalAlpha. I was trying to think how best to incorporate it more.

**Comment by JuliusSweetland** [comment]
Great stuff. I can help later this week.

**Comment by zelmon64** [comment]
@JuliusSweetland I've done everything on your list apart from prompt the user to turn ForceCapsLock on or off because I'm not sure if it's really necessary as I think the key works quite nicely with ForceCapsLock off. 

I've set all the letter and punctuation keys to return to the main layer and the number keys return to the number layer.

Here's a build in case you (or anyone else) want to try it before merging.
https://drive.google.com/open?id=0B7SYYOqNW4FSOTBNcjlYQnlub1U

**Comment by JuliusSweetland** [comment]
Awesome. Thank you. I'll check it out this evening.

File: src/JuliusSweetland.OptiKey/UI/ViewModels/Keyboards/SimplifiedAlpha.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using JuliusSweetland.OptiKey.UI.ViewModels.Keyboards.Base;
+
+namespace JuliusSweetland.OptiKey.UI.ViewModels.Keyboards
+{
+    public class SimplifiedAlpha : BackActionKeyboard, IConversationKeyboard
+    {
+        public SimplifiedAlpha(Action backAction)
+            : base(backAction, simulateKeyStrokes: false, multiKeySelectionSupported: true)
```
Comment (JuliusSweetland) [inline src/JuliusSweetland.OptiKey/UI/ViewModels/Keyboards/SimplifiedAlpha.cs]: multiKeySelectionSupported: false

File: src/JuliusSweetland.OptiKey/UI/ViewModels/Keyboards/SimplifiedConversationAlpha.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using JuliusSweetland.OptiKey.UI.ViewModels.Keyboards.Base;
+
+namespace JuliusSweetland.OptiKey.UI.ViewModels.Keyboards
+{
+    public class SimplifiedConversationAlpha : BackActionKeyboard, IConversationKeyboard
+    {
+        public SimplifiedConversationAlpha(Action backAction)
+            : base(backAction, simulateKeyStrokes: false, multiKeySelectionSupported: true)
```
Comment (JuliusSweetland) [inline src/JuliusSweetland.OptiKey/UI/ViewModels/Keyboards/SimplifiedConversationAlpha.cs]: multiKeySelectionSupported: false

File: src/JuliusSweetland.OptiKey/Enums/Keyboards.cs [hunk 3]
```diff
@@ -18,6 +18,7 @@ public enum Keyboards
         NumericAndSymbols2,
         NumericAndSymbols3,
         PhysicalKeys,
+        SimplifiedAlpha,
```
Comment (JuliusSweetland) [inline src/JuliusSweetland.OptiKey/Enums/Keyboards.cs]: Is this required?

Comment (zelmon64) [inline src/JuliusSweetland.OptiKey/Enums/Keyboards.cs]: No it's not. Thanks for spotting it and the others.

**Comment by JuliusSweetland** [comment]
You're quick! I was leaving a few comments for myself as I take a break to
eat. I'll continue in a bit. Looks really good though. Great quality.

File: src/JuliusSweetland.OptiKey/UI/Views/Keyboards/English/SimplifiedAlpha.xaml [hunk 4]
```diff
@@ -0,0 +1,1285 @@
...
+                                          Text="{x:Static resx:Resources.RE_CALIBRATE}"
+                                          SharedSizeGroup="KeyWithSymbol"
+                                          WidthSpan="2" 
+                                          Value="{x:Static models:KeyValues.CalibrateKey}"/>
+                        </Setter.Value>
+                    </Setter>
+                    <Style.Triggers>
+                        <MultiDataTrigger>
+                            <MultiDataTrigger.Conditions>
+                                <Condition Binding="{Binding Source={x:Static properties:Settings.Default}, Path=ForceCapsLock}" Value="False" />
```
Comment (JuliusSweetland) [inline src/JuliusSweetland.OptiKey/UI/Views/Keyboards/English/SimplifiedAlpha.xaml]: Is this a mistake? I can't see why the Calibrate key's visibility is linked to the ForceCapsLock setting. Don't worry about submitting a commit - I will remove if it's not linked - can you just let me know? This is also the same on the SimplifiedConversationAlpha (on 2 Calibration keys).

Comment (zelmon64) [inline src/JuliusSweetland.OptiKey/UI/Views/Keyboards/English/SimplifiedAlpha.xaml]: It's not a mistake. When ForceCapsLock is disabled the shift key uses some of the space the Calibrate key is in. It's the same for the Conversation Confirm keyboard key.

Comment (JuliusSweetland) [inline src/JuliusSweetland.OptiKey/UI/Views/Keyboards/English/SimplifiedAlpha.xaml]: Ah, thanks for clarifying

File: src/JuliusSweetland.OptiKey/UI/Views/Keyboards/English/SimplifiedConversationAlpha.xaml [hunk 5]
```diff
@@ -0,0 +1,1249 @@
...
+                        Grid.RowSpan="1" Grid.ColumnSpan="1">
+            <ContentControl.Style>
+                <Style TargetType="{x:Type ContentControl}">
+                    <Setter Property="Visibility" Value="Hidden" />
+                    <Setter Property="Content">
+                        <Setter.Value>
+                            <controls:Key SymbolGeometry="{StaticResource CalibrateIcon}"
+                                          Text="{x:Static resx:Resources.RE_CALIBRATE}"
+                                          SharedSizeGroup="KeyWithSymbol"
+                                          WidthSpan="2"
```
Comment (JuliusSweetland) [inline src/JuliusSweetland.OptiKey/UI/Views/Keyboards/English/SimplifiedConversationAlpha.xaml]: Should this key be WidthSpan="2"? The other version of this key actually spans 2 columns, this one only spans 1 column.

File: src/JuliusSweetland.OptiKey/UI/Views/Keyboards/English/SimplifiedConversationAlpha.xaml [hunk 6]
```diff
@@ -0,0 +1,1249 @@
...
+                                <Condition Binding="{Binding Source={x:Static properties:Settings.Default}, Path=ConversationConfirmEnable}" Value="True" />
+                            </MultiDataTrigger.Conditions>
+                            <Setter Property="Visibility" Value="Visible" />
+                        </MultiDataTrigger>
+                    </Style.Triggers>
+                </Style>
+            </ContentControl.Style>
+        </ContentControl>
+
+        <ContentControl Grid.Row="3" Grid.Column="0"
```
Comment (JuliusSweetland) [inline src/JuliusSweetland.OptiKey/UI/Views/Keyboards/English/SimplifiedConversationAlpha.xaml]: This is a duplicate key - removing.

File: src/JuliusSweetland.OptiKey/UI/Views/Keyboards/English/SimplifiedConversationAlpha.xaml (hunk 5 above)
Comment (zelmon64) [inline src/JuliusSweetland.OptiKey/UI/Views/Keyboards/English/SimplifiedConversationAlpha.xaml]: I think you're right and I simply forgot to delete it when I copied it.

**Comment by JuliusSweetland** [comment]
@zelmon64 Merged. An excellent pull request - thank you very much! If there's anything else you would like to work on at any time please let me know.

**Comment by JuliusSweetland** [comment]
@zelmon64 This has been released as part of v2.4.0 - https://github.com/OptiKey/OptiKey/releases/tag/v2.4.0

**Pull request merged**